awx/api/serializers.py

```diff
@@ -3364,6 +3370,70 @@ class Meta:
         fields = ('can_cancel',)
 
 
+class WorkflowApprovalViewSerializer(UnifiedJobSerializer):
+
+    class Meta:
+        model = WorkflowApproval
+        fields = []
+
+
+class WorkflowApprovalSerializer(UnifiedJobSerializer):
+
+    class Meta:
+        model = WorkflowApproval
+        fields = (['*', '-controller_node', '-execution_node',])
+
+    def get_related(self, obj):
+        res = super(WorkflowApprovalSerializer, self).get_related(obj)
+
+        if obj.workflow_approval_template:
+            res['workflow_approval_template'] = self.reverse('api:workflow_approval_template_detail',
+                                                             kwargs={'pk': obj.workflow_approval_template.pk})
+            res['notifications'] = self.reverse('api:workflow_approval_notifications_list', kwargs={'pk': obj.pk})
+        res['approve'] = self.reverse('api:workflow_approval_approve', kwargs={'pk': obj.pk})
+        res['deny'] = self.reverse('api:workflow_approval_deny', kwargs={'pk': obj.pk})
+        return res
```

Comment:
We should add an additional `SerializerMethodField` here for `can_approve_or_deny` e.g.,

```diff
+ can_approve_or_deny = serializers.SerializerMethodField('can_approve_or_deny')
...
+     fields = ['*', '-controller_node', '-execution_node', 'can_approve_or_deny']
...
+ def get_can_approve_or_deny(self, obj):
+    request = self.context.get('request', None)
+    return request.user.can_access(models. WorkflowApproval, 'approve_or_deny', obj)
```

Reply:
Maybe we should add this on the approval list view, too?  @AlanCoding what do you think about that performance-wise?

Reply:
> Maybe we should add this on the approval list serializer, too? @AlanCoding what do you think about that performance-wise?

Before adding any permissions-related values in a list view, we should have a concrete reason driven by UI proposals (because of API performance hit).

I _think_ that is satisfied in this case, but I do not know for sure. I presume that the UI will grey entries in the approval jobs list or somehow show that the user cannot approve them. They might also filter the list. I don't know for certain which.

I do not think the approval JTs would ever reasonably have this property. I used the term "approval job", but the more canonical term might be just "approval".

I'm not as concerned about this as I'd be for other unified job types, because IIRC, this would only get used in the list for approval jobs (which has its own view), and not in the unified jobs list.

Reply:
@AlanCoding yea, I know we've talked about a global notification UI where you can view a list of pending "approvals".  I'd expect we're surface the `can_approve_or_deny` there so that the list could conditionally grey out or display the approve/deny buttons.  In this way, users could *see* what's waiting on approval, but not have active buttons if they're not in the proper role.

Reply:
> Performance-wise, I think this is fine since we're explicitly filtering WorkflowApproval from the global UJ list view.

Agreed.

Any form of prefetching these values for a list view is quite complex, so in the merged version of this feature, the serializer can just call out to the access method.